lib/galaxy/webapps/galaxy/services/wes.py

```diff
@@ -0,0 +1,985 @@
+            state=GALAXY_TO_WES_STATE.get(invocation.state or "", State.UNKNOWN),
+        )
+
+    def cancel_run(
+        self,
+        trans: ProvidesUserContext,
+        run_id: int,
+    ) -> RunId:
+        """Cancel a running workflow.
+
+        Args:
+            trans: Galaxy transaction/context
+            run_id: The WES run ID (Galaxy invocation ID)
+
+        Returns:
+            RunId of the cancelled run
+        """
+        invocation = self._get_invocation(trans, run_id)
+
+        # Request cancellation through Galaxy's workflows manager
+        cancelled_invocation = self._workflows_service._workflows_manager.request_invocation_cancellation(
+            trans, invocation.id
+        )
+        return self._invocation_to_run_id(cancelled_invocation)
+
+    def get_run_tasks(
+        self,
+        trans: ProvidesUserContext,
+        run_id: int,
+        page_size: int = 10,
+        page_token: Optional[str] = None,
+    ) -> TaskListResponse:
+        """Get paginated list of tasks for a workflow run.
+
+        Args:
+            trans: Galaxy transaction/context
+            run_id: The WES run ID (Galaxy invocation ID)
+            page_size: Number of tasks per page (default 10, max 100)
+            page_token: Token for pagination (base64-encoded offset)
+
+        Returns:
+            TaskListResponse with paginated tasks
+        """
+        invocation = self._get_invocation(trans, run_id)
+
+        # Get all task logs
+        all_task_logs = self._build_task_logs(trans, invocation)
+
+        # Decode page token to offset
+        offset = _decode_page_token(page_token)
+
+        # Apply pagination
+        end_offset = offset + page_size
+        paginated_logs = all_task_logs[offset:end_offset]
+
+        # Generate next page token if more results exist
+        next_page_token = None
+        if end_offset < len(all_task_logs):
+            next_page_token = _encode_page_token(end_offset)
+
+        return TaskListResponse(
+            task_logs=paginated_logs if paginated_logs else None,
+            next_page_token=next_page_token,
+        )
+
+    def get_run_task(
+        self,
+        trans: ProvidesUserContext,
+        run_id: int,
+        task_id: str,
+    ) -> TaskLog:
+        """Get details for a specific task in a workflow run.
+
+        Args:
+            trans: Galaxy transaction/context
+            run_id: The WES run ID (Galaxy invocation ID)
+            task_id: The task ID (step ID or step.job format)
+
+        Returns:
+            TaskLog object for the specified task
+
+        Raises:
+            exceptions.ObjectNotFound: If task not found
+        """
+        invocation = self._get_invocation(trans, run_id)
+
+        # Get all task logs
+        all_task_logs = self._build_task_logs(trans, invocation)
+
+        # Find matching task by ID
+        for task_log in all_task_logs:
+            if task_log.id == task_id:
```

Comment:
As above, loading all tasks into memory to find one by id is suboptimal.

Reply:
Also - gone.